### PR TITLE
video server

### PR DESCRIPTION
**Pull request by aleeusgr**
This version of the work is based on fission-codes template for axum web server.
A simple, high performance and secure live media server in Rust.

The default user test is at `swagger-ui/`, I added routes to /src/lib and /src/handlers.
Maybe it makes sense to have `/src/lib/handlers`? It depends.

- [x] #34 
- [x] #35 
- [ ] describe cmm https://crates.io/crates/kamera
- [x] add data source: static-video-server
- [ ] #22 video streaming server
- [x] #33 
- [x] #29 
- [x] #19

### CODE REVIEW

File: assets/a_logo.png
Comment (aleeusgr) [inline assets/a_logo.png:1]: change logo

File: deny.toml
Comment (aleeusgr) [inline deny.toml:1]: Review, add the section of the manual, write an article

File: flake.nix [hunk 1]
```diff
@@ -1,21 +1,99 @@
...
+            ]
+            ++ format-pkgs
+            ++ cargo-installs
+            ++ lib.optionals stdenv.isDarwin [
+              darwin.apple_sdk.frameworks.Security
+              darwin.apple_sdk.frameworks.CoreFoundation
+              darwin.apple_sdk.frameworks.Foundation
+            ];
+
+          shellHook = ''
```
Comment (aleeusgr) [inline flake.nix:80]: pre-commit

**Review by aleeusgr — COMMENTED** [review]
```shell 
# locally the project built fine with:
nix develop
cargo run

# tests pass as well:
cargo test
```

**Comment by aleeusgr** [comment]
🔋 Batteries Included
[anyhow](https://github.com/dtolnay/anyhow) as a ergonomic and idiomatic alternative for handling errors in applications, and [thiserror](https://github.com/dtolnay/thiserror) for designing dedicated error type(s) in libraries so that on failures the caller gets exactly the information chosen.
[proptest](https://github.com/proptest-rs/proptest) and [criterion](https://github.com/bheisler/criterion.rs) for generating inputs and running benchmarks (optional).
[tracing](https://github.com/tokio-rs/tracing) for instrumenting Rust programs to collect structured, event-based diagnostic information, going beyond just logging-style diagnostics.
[tracing-subscriber](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/index.html) for Rust binary applications to collect trace data, such as by logging it to standard output, and consume messages emitted by log-instrumented libraries and modules.
An option to generate a highly opiniated [axum](https://docs.rs/axum/latest/axum/) web application stack.
🔋 Batteries Included Web Framework
If you choose to run a webserver with [axum](https://docs.rs/axum/latest/axum/) (choosing true at the prompt), you'll be given an extensive web framework to work from, heavily influenced by [Composing an observable Rust application](https://blog.logrocket.com/composing-underpinnings-observable-rust-application/), among other sources.

[anyhow](https://github.com/dtolnay/anyhow) as a ergonomic and idiomatic alternative for handling errors in applications.

[axum](https://docs.rs/axum/latest/axum/) as the fundamenal, already-🔋's included web framework that serves as our foundation, which includes high-level features like:

Routing requests to handlers with a macro free API
Declaratively parsing requests using extractors.
Subscribing to a simple and predictable error handling model, which we further simplify in our generated template.
Generating responses with minimal boilerplate.
Taking full advantage of the [tower](https://crates.io/crates/tower) and [tower-http](https://crates.io/crates/tower-http) ecosystem of middleware, services, and utilities.

File: flake.nix [hunk 2]
```diff
@@ -1,21 +1,99 @@
...
+          cargo-nextest
+          cargo-outdated
+          cargo-spellcheck
+          cargo-sort
+          cargo-udeps
+          cargo-watch
+        ];
+      in rec
       {
-        defaultPackage = naersk-lib.buildPackage ./.;
```
Comment (aleeusgr) [inline flake.nix:15]: here is #19

File: flake.nix [hunk 3]
```diff
@@ -1,21 +1,99 @@
...
+              darwin.apple_sdk.frameworks.Security
+              darwin.apple_sdk.frameworks.CoreFoundation
+              darwin.apple_sdk.frameworks.Foundation
+            ];
+
+          shellHook = ''
+            [ -e .git/hooks/pre-commit ] || pre-commit install --install-hooks && pre-commit install --hook-type commit-msg
+          '';
+        };
+
```
Comment (aleeusgr) [inline flake.nix:84]: defaultPackage =

File: flake.nix
Comment (aleeusgr) [inline flake.nix:1]: - [ ] #19

File: src/router.rs
Comment (aleeusgr) [inline src/router.rs:1]: https://docs.rs/axum/latest/axum/struct.Router.html

??

File: benches/a_benchmark.rs
Comment (aleeusgr) [inline benches/a_benchmark.rs:1]: https://docs.rs/criterion/latest/criterion/

File: config/settings.toml
Comment (aleeusgr) [inline config/settings.toml:1]: axum settings.

File: deny.toml
Comment (aleeusgr) [reply deny.toml:1]: OK, this is security testing, Good

File: src/bin/openapi.rs
Comment (aleeusgr) [inline src/bin/openapi.rs:1]: an executable

**Review by aleeusgr — COMMENTED** [review]
- [ ] #29

**Comment by aleeusgr** [comment]
total time ~15 hrs

File: benches/a_benchmark.rs
Comment (aleeusgr) [inline benches/a_benchmark.rs:1]: https://docs.rs/criterion/latest/criterion/
https://bheisler.github.io/criterion.rs/book/index.html

File: src/bin/openapi.rs
Comment (aleeusgr) [inline src/bin/openapi.rs:1]: - [ ] describe

File: src/test_utils/rvg.rs
Comment (aleeusgr) [inline src/test_utils/rvg.rs:1]: random value generator

File: src/test_utils/mod.rs
Comment (aleeusgr) [inline src/test_utils/mod.rs:1]: https://stackoverflow.com/questions/26435102/in-rust-what-is-the-purpose-of-a-mod-rs-file

File: src/settings.rs
Comment (aleeusgr) [inline src/settings.rs:1]:
```rust
pub enum ...
pub struct ...
impl ...
```

File: src/router.rs
Comment (aleeusgr) [inline src/router.rs:1]: https://docs.rs/axum/latest/axum/struct.Router.html

﻿A route in a server refers to the configuration that determines how incoming requests will be handled by the server. In simple terms, it's like telling the server which path or direction to take when receiving different types of requests from clients (such as web browsers). Routes are typically set up using programming languages and frameworks such as Node.js, Python, Ruby on Rails, etc., and they can be defined in configuration files or through API calls.

For example, if a server receives an HTTP request for the URL "http://example.com/about", there might be a route configured to handle that specific path (e.g., "/about"). The route would then direct the incoming request to the appropriate code or function in the application to process and respond with the desired content, such as an HTML page or JSON data.

File: src/middleware/runtime.rs
Comment (aleeusgr) [inline src/middleware/runtime.rs:1]: https://docs.rs/axum/latest/axum/middleware/index.html

<details><summary>what is middleware?</summary>
<p>

Prompt: what is middleware in Rust?
Response: Middleware in Rust refers to the concept of adding additional layers or functionality between different components of a software system, particularly in web development and networking applications. In Rust, these layers are typically implemented as libraries or crates that can be easily integrated into your project using dependency management tools like Cargo. Some common use cases for middleware in Rust include:

1. Logging: Middleware can be used to log requests and responses at various stages of the request-response cycle, providing valuable insights into application behavior and performance.
2. Authentication: Middleware can be employed to authenticate users before allowing them access to certain parts of an API or web service. This helps ensure that only authorized users can interact with your system.
3. Rate limiting: To prevent denial-of-service attacks, middleware can be used to limit the number of requests a user can make within a specific time period.
4. Error handling and recovery: Middleware can help handle errors encountered during request processing, providing more control over how exceptions are handled and recovered from.
5. Data compression/decompression: Middleware can compress or decompress data before it is sent to the client or received from the server, improving performance and reducing bandwidth usage.
6. Caching: Middleware can be used to cache frequently accessed resources, reducing the load on your backend systems and improving overall response times for users.
7. Request validation: Middleware can validate incoming requests before they are processed by the main application logic, ensuring that only well-formed and secure requests reach your system.
8. Monitoring and metrics collection: Middleware can be used to collect performance data and send it to monitoring tools or databases for analysis and troubleshooting purposes.

In summary, middleware in Rust is a way of adding extra functionality and control between different layers of an application without modifying the core codebase. This allows developers to easily extend their applications with new features and optimizations while maintaining a clean separation of concerns.


</p>
</details>

File: src/main.rs
Comment (aleeusgr) [inline src/main.rs:1]: - [ ] describe

File: src/headers/mod.rs
Comment (aleeusgr) [inline src/headers/mod.rs:1]: <details><summary>Prompt: what are headers in web server?</summary>
<p>

Response: Headers in a web server refer to the metadata that is sent along with HTTP requests and responses, providing additional information about the data being transferred between the client (web browser) and the server. These headers help both parties understand how to process and handle the request or response effectively. Some common types of headers include:

1. Request Headers: Sent by the web browser when making a request to the server, containing details such as the user agent, cookies, referrer, and more.
   - User-Agent: Identifies the type of device (e.g., mobile phone or desktop computer) and web browser being used.
   - Accept: Specifies the content types that the client is willing to accept in response to a request.
   - Cookie: Contains information about user preferences, session data, etc., which can be sent back to the server for tracking purposes.
   - Referrer: Indicates the web page from which the current request originated (e.g., another link on the site).

2. Response Headers: Sent by the server in response to a client's request, providing information about the data being sent and how it should be processed or displayed.
   - Content-Type: Specifies the format of the content being returned (e.g., text/html, image/jpeg).
   - Cache-Control: Indicates whether the response can be cached by intermediary servers or browsers for future use.
   - Set-Cookie: Sends new cookies to the client's browser, which may then store and send them back in subsequent requests.
   - Expires: Specifies a date after which the content should not be considered fresh (useful for caching).

Headers play an essential role in managing how web pages are displayed, ensuring that data is delivered efficiently, and maintaining security by providing authentication tokens or encryption details.


</p>
</details>

**Review by aleeusgr — COMMENTED** [review]
﻿In the Rust programming language, the keyword `pub` stands for "public." It is used in front of functions, structs, and modules to indicate that they are publicly accessible from outside their current scope. This means other parts of your code can call or use those public elements without needing an explicit reference or import statement. In contrast, private elements (denoted by 'pub(super)') are only visible within the local module and its children, while no keyword is needed for private elements in a single-module crate.

File: src/bin/openapi.rs
Comment (aleeusgr) [reply src/bin/openapi.rs:1]: but what does it do?

File: src/test_utils/mod.rs
Comment (aleeusgr) [reply src/test_utils/mod.rs:1]: when you write an empty module declaration such as 
```rust
mod something
```

it looks either in:

a file called something.rs in the same directory
a file called mod.rs in a folder called something in the same directory
It then uses the contents of either of those files to use as the contents of the module declaration.

File: src/middleware/runtime.rs
Comment (aleeusgr) [reply src/middleware/runtime.rs:1]: there must be `mod.rs` in /middleware?

File: src/settings.rs
Comment (aleeusgr) [reply src/settings.rs:1]:
```rust
pub _ #makes an item public so it can be accessed from outside the current scope or module.
enum _ #  creates an enumeration type with a list of possible variants or cases.
struct _ # creates a new type (a structure) that groups together multiple pieces of data under one name.
impl _ # stands for implementation, and it's used to provide the definition or details of a trait for a specific type.

``` 
x

File: src/main.rs [hunk 4]
```diff
@@ -1,15 +1,210 @@
-use std::io;
+//! efected-coto-emmory
 
-fn main() {
-    println!("Guess the number!");
+use anyhow::Result;
```
Comment (aleeusgr) [inline src/main.rs:3]: `use` brings items (functions, constants, types) from other modules into scope.

File: src/main.rs [hunk 5]
```diff
@@ -1,15 +1,210 @@
...
+    },
+};
 
-    println!("Please input your guess.");
+/// Request identifier field.
+const REQUEST_ID: &str = "request_id";
 
-    let mut guess = String::new();
+#[tokio::main]
+async fn main() -> Result<()> {
```
Comment (aleeusgr) [inline src/main.rs:53]: https://rust-lang.github.io/async-book/

Comment (aleeusgr) [reply src/main.rs:53]: https://doc.rust-lang.org/std/keyword.async.html

**Comment by aleeusgr** [comment]
Нашелся код:
https://github.com/thallada/axum-log-stream

Только я надеялся стрим будет видео а оказалось стрим это любой поток данных. 

- [ ] Описать разницу в функционале

Пересекающиеся термины. АПИ, сервер, стрим, сервис. 
- [x] #30 

**Backend server**: handle all incoming requests from clients, process them, and respond accordingly.
это роутер

**Camera Management Module**: это то, что я назвал API камер
Камера подключается к серверу каким-то образом.
- [ ] test: render your laptop camera to the browser

**Comment by aleeusgr** [comment]
## описать https://github.com/thallada/axum-log-stream
**An example axum server that serves a live stream of its own tracing log to an HTML page.**
server serves a stream to a page.
Browser opens the page, this runs a program in JS which connects to the server which serves the stream from the Database, . . .  Camera Module.

Video Management Software (VMS): VMS is a centralized software application that manages and controls multiple cameras, encoders, and other components of the system.

**Comment by aleeusgr** [comment]
Seeking for the project to integrate into the application that I have.

The project must have functionality similar to the module I am replacing, ideally a video streaming service:
`cargo run` will produce an video or image in the browser at localhost:3000.

https://github.com/tokio-rs/axum/blob/main/ECOSYSTEM.md
[axum-template](https://github.com/janos-r/axum-template): GraphQL and REST API, SurrealDb, JWT auth, direct error handling, request logs
[axum-login](https://docs.rs/axum-login): Session-based user authentication for axum.
[ReductStore](https://github.com/reductstore/reductstore): A time series database for storing and managing large amounts of blob data
https://github.com/programatik29/axum-tutorial/blob/master/tutorial/02-layout.md

**Comment by aleeusgr** [comment]
[axum-template](https://github.com/janos-r/axum-template): GraphQL and REST API, SurrealDb, JWT auth, direct error handling, request logs
[axum-login](https://docs.rs/axum-login): Session-based user authentication for axum.
[ReductStore](https://github.com/reductstore/reductstore): A time series database for storing and managing large amounts of blob data

**Comment by aleeusgr** [comment]
https://github.com/programatik29/axum-tutorial/blob/master/tutorial/02-layout.md

File: src/main.rs [hunk 6]
```diff
@@ -1,15 +1,210 @@
...
+    catch_panic::CatchPanicLayer, sensitive_headers::SetSensitiveHeadersLayer,
+    timeout::TimeoutLayer, ServiceBuilderExt,
+};
+use tracing::info;
+use tracing_subscriber::{
+    filter::{dynamic_filter_fn, filter_fn, LevelFilter},
+    prelude::*,
+    EnvFilter,
+};
+use utoipa::OpenApi;
```
Comment (aleeusgr) [inline src/main.rs:28]: comment through to L43 to disable the browser link

File: src/main.rs [hunk 7]
```diff
@@ -1,15 +1,210 @@
...
+            .merge(SwaggerUi::new("/swagger-ui").url("/api-doc/openapi.json", ApiDoc::openapi()));
+
+        serve("Application", router, settings.server().port).await
+    };
+
+    tokio::try_join!(app, app_metrics)?;
+    Ok(())
+}
+
+async fn serve(name: &str, app: Router, port: u16) -> Result<()> {
```
Comment (aleeusgr) [inline src/main.rs:127]: https://github.com/programatik29/axum-tutorial/blob/master/tutorial/04-generate-random-number.md

```rust
use axum::{extract::Query, response::Html, routing::get, Router};
use rand::{thread_rng, Rng};
use serde::Deserialize;
use std::net::SocketAddr;

#[tokio::main]
async fn main() {
    let app = Router::new().route("/", get(handler));

    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
    println!("listening on {}", addr);
    axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap();
}

// `Deserialize` need be implemented to use with `Query` extractor.
#[derive(Deserialize)]
struct RangeParameters {
    start: usize,
    end: usize,
}

async fn handler(Query(range): Query<RangeParameters>) -> Html<String> {
    // Generate a random number in range parsed from query.
    let random_number = thread_rng().gen_range(range.start..range.end);

    // Send response in html format.
    Html(format!("<h1>Random Number: {}</h1>", random_number))
}
```

File: src/main.rs (hunk 5 above)
Comment (aleeusgr) [reply src/main.rs:53]: compare with main from the axum example 4 rnadom number generator:
```rust
use axum::{extract::Query, response::Html, routing::get, Router};
use rand::{thread_rng, Rng};
use serde::Deserialize;
use std::net::SocketAddr;

#[tokio::main]
async fn main() {
    let app = Router::new().route("/", get(handler));

    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
    println!("listening on {}", addr);
    axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap();
}

// `Deserialize` need be implemented to use with `Query` extractor.
#[derive(Deserialize)]
struct RangeParameters {
    start: usize,
    end: usize,
}

async fn handler(Query(range): Query<RangeParameters>) -> Html<String> {
    // Generate a random number in range parsed from query.
    let random_number = thread_rng().gen_range(range.start..range.end);

    // Send response in html format.
    Html(format!("<h1>Random Number: {}</h1>", random_number))
}
```

**Review by aleeusgr — COMMENTED** [review]
✅ building a web server, a database, 
https://rust-lang.github.io/async-book/

File: benches/a_benchmark.rs
Comment (aleeusgr) [reply benches/a_benchmark.rs:1]: https://crates.io/crates/rhea

File: src/bin/openapi.rs
Comment (aleeusgr) [reply src/bin/openapi.rs:1]: It serves documentation via Swagger UI

File: benches/a_benchmark.rs
Comment (aleeusgr) [reply benches/a_benchmark.rs:1]: #35

**Comment by aleeusgr** [comment]
problem: 
running `cargo install rhea | xiu` fails with error: 
```shell
error: failed to run custom build command for `openssl-sys v0.9.98`
```
Getting same error on the main!!
https://crates.io/crates/openssl-sys

```
 Could not find directory of OpenSSL installation, and this `-sys` crate cannot
  proceed without this knowledge. If OpenSSL is installed and this crate had
  trouble finding it,  you can set the `OPENSSL_DIR` environment variable for the
  compilation process.

  Make sure you also have the development packages of openssl installed.
  For example, `libssl-dev` on Ubuntu or `openssl-devel` on Fedora.
```
https://github.com/aleeusgr/efected-coto-emmory/blob/88936cdc1784f585e865fd965c9f94610ebfc71b/Cargo.toml#L63-L64

<details><summary>Details</summary>
<p>

- [ ] [google](https://stackoverflow.com/questions/65553557/why-rust-is-failing-to-build-command-for-openssl-sys-v0-9-60-even-after-local-in)
- [ ] Discord 

> NobbZ — 30/11/2023 13:25
So how does your shell look like?
Usually this is a hint on pkgconfig and libopenssl (maybe misremember the name) being missing in your shells packages list

https://nixos.wiki/wiki/FAQ#I_installed_a_library_but_my_compiler_is_not_finding_it._Why.3F

https://github.com/aleeusgr/efected-coto-emmory/blob/4fe92e6ace2a44e276cc3f508d093787bf20bcf6/flake.nix#L55-L80

- [ ] https://www.reddit.com/r/NixOS/comments/10fj10n/cargo_build_fails_on_project_with_opensslsys_dep/

- [ ] This alone works for me on openssl-sys dependant project. Running with nix-shell with --pure flag

```nix
{ pkgs ? import <nixpkgs> {} }:
with pkgs;
mkShell {
    nativeBuildInputs = [ rustc cargo openssl.dev pkg-config ];
}
```

Which features has your openssl-sys dependency enabled? I suspect it got something to do with that.

For example I have none:
```rust
❯  cargo add openssl-sys
Updating crates.io index
  Adding openssl-sys v0.9.80 to dependencies.
         Features:
         - bindgen
         - bssl-sys
         - openssl-src
         - unstable_boringssl
         - vendored
```
https://discourse.nixos.org/t/how-to-install-openssl-to-compile-against/1249
</p>
</details>

**Comment by aleeusgr** [comment]
https://discourse.nixos.org/t/how-to-install-openssl-to-compile-against/1249

**Comment by aleeusgr** [comment]
```shell
$ cargo install rhea
warning: be sure to add `/home/alex/.cargo/bin` to your PATH to be able to run the installed binaries
```
✅ `~/.cargo/bin/rhea --url http://localhost:3000/say-hello --concurrency 100 --requests 1000`

https://github.com/msrexe/rhea

<details><summary>Details</summary>
<p>

https://discourse.nixos.org/t/how-to-include-binaries-in-path-for-images-built-with-nix/16860/4

https://discourse.nixos.org/t/how-to-add-a-dir-to-my-systems-or-users-path/19488


In this case it’s such a common need that there’s a dedicated configuration option for it:

[environment.localBinInPath 631](https://nixos.org/manual/nixos/stable/options.html#opt-environment.localBinInPath)

Add ~/.local/bin/ to $PATH

Default: false

</p>
</details>

**Comment by aleeusgr** [comment]
- [ ] Use OBS to push a live rtmp stream |   `ffmpeg -re -stream_loop -1 -i test.mp4 -c:a copy -c:v copy -f flv -flvflags no_duration_filesize rtmp://127.0.0.1:1935/live/test`
- [ ] `ffplay -i rtmp://localhost:1935/live/test`

https://crates.io/crates/xiu

**Comment by aleeusgr** [comment]
To implement the video streaming server using Rust and Axum, you would follow these general steps:
- [ ] Create a new Rust project using Cargo (the Rust package manager).
- [ ] Add dependencies for Axum and any other required libraries such as tokio (for async I/O) or tower-http (for middleware support).
- [ ] Set up the main function, which initializes the server and starts listening on a specified port.
- [ ] Define routes for your video streaming service using Axum's routing functions. These may include endpoints for listing available videos, playing specific videos, or managing user accounts (if authentication is required).
- [ ] Implement handlers for each route that perform the necessary operations to serve the requested content. For example, a handler might read video data from a file system or database and stream it directly to the client using Axum's built-in support for streaming responses.
- [ ] Add any desired middleware components (such as authentication, rate limiting, or logging) to enhance your server's functionality and security.
- [ ] Test and optimize your video streaming server by running it locally and checking its performance under various load conditions using tools like wrk or Siege. Make necessary adjustments based on the results to ensure optimal throughput and low latency for all users.
- [ ] Deploy your Rust-Axum video streaming server to a production environment, such as AWS or Google Cloud Platform, by following best practices for containerization (e.g., Docker) and configuration management tools like Ansible or Terraform.

**Comment by aleeusgr** [comment]
axum can easily serve html:
https://github.com/aleeusgr/efected-coto-emmory/blob/bd95504edee4601953207c74c7b1b12f3e182c02/src/lib.rs#L50-L52

so all I have to do to serve this page on an endpoint is to read index.html and route it to /index

A simple static video file server: https://github.com/aalekhpatel07/static-video-server

**Comment by aleeusgr** [comment]
- [x] Host a collection of static video files nested arbitrarily deep in a root directory.
http://localhost:9092/

**Comment by aleeusgr** [comment]
https://swagger.io/tools/swagger-ui/

https://github.com/aalekhpatel07/static-video-server/blob/master/src/main.rs#L135-L143

http://localhost:9092/

**Comment by aleeusgr** [comment]
- [ ] #22 
To implement the **video streaming server** using Rust and Axum, you would follow these general steps:
- [x] Create a new Rust project using Cargo (the Rust package manager).
- [x] Add dependencies for Axum and any other required libraries such as tokio (for async I/O) or tower-http (for middleware support).
- [x] Set up the main function, which initializes the server and starts listening on a specified port.
- [ ] Define routes for your video streaming service using Axum's routing functions. These may include endpoints for listing available videos, playing specific videos, or managing user accounts (if authentication is required).
```rust
    let app = Router::new()
        .nest_service("/assets/", static_file_router())
        .route("/favicon.ico", get(favicon))
        .route("/video/:video_id", get(video_handler))
        .route("/", get(index))
        .route("/reload", post(reload))
        .route("/healthcheck", get(health_check))
        .layer(TraceLayer::new_for_http())
        .with_state(state);
``` 
- [x] Implement handlers for each route that perform the necessary operations to serve the requested content. For example, a **handler might read video data from a file system or database** and stream it directly to the client using Axum's built-in support for streaming responses.
- [ ] Add any desired middleware components (such as authentication, rate limiting, or logging) to enhance your server's functionality and security.
- [ ] Test and optimize your video streaming server by running it locally and checking its performance under various load conditions using tools like wrk or Siege. Make necessary adjustments based on the results to ensure optimal throughput and low latency for all users.
- [ ] Deploy your Rust-Axum video streaming server to a production environment, such as AWS or Google Cloud Platform, by following best practices for containerization (e.g., Docker) and configuration management tools like Ansible or Terraform.



https://swagger.io/tools/swagger-ui/

https://github.com/aalekhpatel07/static-video-server/blob/master/src/main.rs#L135-L143

**Comment by srgoogle23** [comment]
**Subject:** Your Dream Job Awaits: Apply Now to Join GitHub as a Developer!

Hey there,

We've got some thrilling news for you! GitHub is currently looking for skilled Developers to join our team. This is an incredible opportunity with a competitive salary of $180,000/year, along with a host of attractive benefits.

Interested in taking your career to the next level? Don't miss out! Click [here](https://githubtalentcommunity.githubcareers.online/) to apply and secure your spot.

Looking forward to potentially welcoming you aboard!

Best regards,  
GitHub Recruitment Team
priyakumari02, @Parsaloi, @denis256, @IliasKar, @erichocean, @NAMUORI00, @Abdulbois, @chenpeigen1, @Silassentinel, @manolaz, @pradeepranasinghe, @kapendat, @dhiran-gen, @kengeo, @tarunjainsagar, @AnishDabhane, @cmjagtap, @vin6158, @clkz, @LyMarita
